plugins/experimental/wasm/wasm_main.cc

```diff
@@ -46,6 +46,250 @@ struct WasmInstanceConfig {
+    toread        = TSVIONTodoGet(input_vio);
+
+    if (toread <= input_avail) { // upstream finished
+      eos = 1;
+    } else {
+      eos = 0;
+    }
+  } else {
+    input_avail   = 0;
+    upstream_done = 0;
+    toread        = 0;
+    eos           = 1;
+  }
+
+  if (input_avail > 0) {
+    // move to the reserved.buffer
+    TSIOBufferCopy(ti->reserved_buffer, input_reader, input_avail, 0);
+
+    // reset input
+    TSIOBufferReaderConsume(input_reader, input_avail);
+    TSVIONDoneSet(input_vio, upstream_done + input_avail);
+  }
+
+  write_down = 0;
+  if (empty_input == 0) {
+    towrite = TSIOBufferReaderAvail(ti->reserved_reader);
+  } else {
+    towrite = 0;
+  }
+
+  do {
+    TSDebug(WASM_DEBUG_TAG, "[%s] inside transform handler loop", __FUNCTION__);
+    proxy_wasm::FilterDataStatus status = proxy_wasm::FilterDataStatus::Continue;
+
+    if (towrite == 0 && empty_input == 0) {
+      break;
+    }
+
+    TSDebug(WASM_DEBUG_TAG, "[%s] retrieving text and calling the wasm handler function", __FUNCTION__);
+    if (empty_input == 0) {
+      blk   = TSIOBufferReaderStart(ti->reserved_reader);
+      start = TSIOBufferBlockReadStart(blk, ti->reserved_reader, &blk_len);
+
+      int size = 0;
+      if (towrite > blk_len) {
+        c->setTransformResult(start, blk_len);
+        towrite -= blk_len;
+        TSIOBufferReaderConsume(ti->reserved_reader, blk_len);
+        size = blk_len;
+      } else {
+        c->setTransformResult(start, towrite);
+        TSIOBufferReaderConsume(ti->reserved_reader, towrite);
+        size    = towrite;
+        towrite = 0;
+      }
+
+      if (!towrite && eos) {
+        if (ti->request) {
+          status = c->onRequestBody(size, true);
+        } else {
+          status = c->onResponseBody(size, true);
+        }
+      } else {
+        if (ti->request) {
+          status = c->onRequestBody(size, false);
+        } else {
+          status = c->onResponseBody(size, false);
+        }
+      }
+    } else {
+      c->setTransformResult(nullptr, 0);
+      if (ti->request) {
+        status = c->onRequestBody(0, true);
+      } else {
+        status = c->onResponseBody(0, true);
+      }
+    }
+
+    TSDebug(WASM_DEBUG_TAG, "[%s] retrieving returns from wasm handler function and pass back to ATS", __FUNCTION__);
+    if ((status == proxy_wasm::FilterDataStatus::Continue) ||
+        ((status == proxy_wasm::FilterDataStatus::StopIterationAndBuffer ||
+          status == proxy_wasm::FilterDataStatus::StopIterationAndWatermark) &&
+         eos && !towrite)) {
+      res = c->getTransformResult(&res_len);
+
+      if (res && res_len > 0) {
+        if (!ti->output_vio) {
+          if (eos && !towrite) {
+            ti->output_vio = TSVConnWrite(output_conn, contp, ti->output_reader, res_len); // HttpSM go on
+          } else {
+            ti->output_vio = TSVConnWrite(output_conn, contp, ti->output_reader, ti->downstream_bytes); // HttpSM go on
```

Comment:
why would this be different? only `transform_result_` gets written right so shouldn't the number of bytes written always be `res_len`?

Reply:
That is probably right. Will try this out.

Reply:
Actually there is a check on line 197 for output_vio. 
so for a long response that comes in multiple block, we will issue TSVConnWrite with INT64_MAX once. For a short response that comes in one single block, we will issue TSVConnWrite with the exact size. 
That's what the code is trying to do.

Reply:
oh for a large response with multiple blocks, it will iterate and keep writing so you are only calling TSVConnWrite to initialize for all of the bytes. got it, thanks for clarifying